File: sdk-ts/src/authentication/device_mode.ts
import { saveCredentials } from "./credentials";
import { Credentials } from "./types";

interface DeviceLogin {
  client_id: string;
  scope: string;
}

interface DeviceLoginResponse {
  client_id: string;
  device_code: string;
  user_code: string;
  expires_in: number;
  interval: number;
  verification_uri: string;
  verification_uri_complete: string;
}

interface DeviceLoginFinalizeRequest {
  grant_type: string;
  client_id: string;
  device_code: string;
}

interface DeviceLoginFinalizeResponse {
  access_token: string;
  expires_in: number;
  refresh_token: string;
  token_type: string;
}

export class BearerToken {
  private credentials: Credentials;
  private workspace_name: string;
  private base_url: string;

  constructor(
    credentials: Credentials,
    workspace_name: string,
    base_url: string
  ) {
    this.credentials = credentials;
    this.workspace_name = workspace_name;
    this.base_url = base_url;
  }

  async getHeaders(): Promise<Record<string, string>> {
    const err = await this.refreshIfNeeded();
    if (err) {
      throw err;
    }
    return {
      "X-Beamlit-Authorization": `Bearer ${this.credentials.access_token}`,
      "X-Beamlit-Workspace": this.workspace_name,
    };
  }

  private async refreshIfNeeded(): Promise<null> {
    // Need to refresh token if expires in less than 10 minutes
    const parts = this.credentials.access_token?.split(".") || [];
    if (parts.length !== 3) {
      throw new Error("Invalid JWT token format");
    }

    try {
      const claims_bytes = Buffer.from(parts[1], "base64url");
      const claims = JSON.parse(claims_bytes.toString());

      // Refresh if token expires in less than 10 minutes
      if (Date.now() / 1000 + 10 * 60 > claims.exp) {
        return await this.doRefresh();
      }
    } catch (e) {
      throw new Error(`Failed to decode/parse JWT claims: ${e}`);
    }

    return null;
  }

  intercept(req: Request): void {
    const err = this.refreshIfNeeded();
    if (err) {
      throw err;
    }

    req.headers.set(
      "X-Beamlit-Authorization",
      `Bearer ${this.credentials.access_token}`
    );
    req.headers.set("X-Beamlit-Workspace", this.workspace_name);
  }

  private async doRefresh(): Promise<null> {
    if (!this.credentials.refresh_token) {
      throw new Error("No refresh token to refresh");
    }

    const url = `${this.base_url}/oauth/token`;
    const refresh_data = {
      grant_type: "refresh_token",
      refresh_token: this.credentials.refresh_token,
      device_code: this.credentials.device_code,
      client_id: "beamlit",
    };

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(refresh_data),
      });

      const finalize_response =
        (await response.json()) as DeviceLoginFinalizeResponse;

      if (!finalize_response.refresh_token) {
        finalize_response.refresh_token = this.credentials.refresh_token;
      }

      const creds: Credentials = {
        access_token: finalize_response.access_token,
        refresh_token: finalize_response.refresh_token,
        expires_in: finalize_response.expires_in,
        device_code: this.credentials.device_code,
        apiKey: "",
        client_credentials: this.credentials.client_credentials,
      };

      this.credentials = creds;
      saveCredentials(this.workspace_name, creds);
      return null;
    } catch (e) {
      throw new Error(`Failed to refresh token: ${e}`);
    }
  }
}

File: sdk/device_mode.go
package sdk

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type DeviceLogin struct {
	ClientID string `json:"client_id"`
	Scope    string `json:"scope"`
}

type DeviceLoginResponse struct {
	ClientID                string `json:"client_id"`
	DeviceCode              string `json:"device_code"`
	UserCode                string `json:"user_code"`
	ExpiresIn               int    `json:"expires_in"`
	Interval                int    `json:"interval"`
	VerificationURI         string `json:"verification_uri"`
	VerificationURIComplete string `json:"verification_uri_complete"`
}
type DeviceLoginFinalizeRequest struct {
	GrantType  string `json:"grant_type"`
	ClientID   string `json:"client_id"`
	DeviceCode string `json:"device_code"`
}

type DeviceLoginFinalizeResponse struct {
	AccessToken  string `json:"access_token"`
	ExpiresIn    int    `json:"expires_in"`
	RefreshToken string `json:"refresh_token"`
	TokenType    string `json:"token_type"`
}

type BearerToken struct {
	credentials   Credentials
	workspaceName string
	baseUrl       string
}

func NewBearerTokenProvider(credentials Credentials, workspaceName string, baseUrl string) *BearerToken {
	return &BearerToken{credentials: credentials, workspaceName: workspaceName, baseUrl: baseUrl}
}

func (s *BearerToken) GetCredentials() Credentials {
	return s.credentials
}

func (s *BearerToken) RefreshIfNeeded() error {
	// Need to refresh the token if access token expires in less than 10 minutes
	parts := strings.Split(s.credentials.AccessToken, ".")
	if len(parts) != 3 {
		return fmt.Errorf("invalid JWT token format")
	}

	claimsBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return fmt.Errorf("failed to decode JWT claims: %v", err)
	}

	var claims map[string]interface{}
	if err := json.Unmarshal(claimsBytes, &claims); err != nil {
		return fmt.Errorf("failed to parse JWT claims: %v", err)
	}

	expTime := time.Unix(int64(claims["exp"].(float64)), 0)
	// Refresh if token expires in less than 10 minutes
	if time.Now().Add(10 * time.Minute).After(expTime) {
		return s.DoRefresh()
	}

	return nil
}

func (s *BearerToken) Intercept(ctx context.Context, req *http.Request) error {
	if err := s.RefreshIfNeeded(); err != nil {
		return err
	}

	req.Header.Set("X-Blaxel-Authorization", fmt.Sprintf("Bearer %s", s.credentials.AccessToken))
	req.Header.Set("X-Blaxel-Workspace", s.workspaceName)
	return nil
}

func (s *BearerToken) DoRefresh() error {
	if s.credentials.RefreshToken == "" {
		return fmt.Errorf("no refresh token to refresh")
	}
	url := s.baseUrl + "/oauth/token"
	refreshData := map[string]string{
		"grant_type":    "refresh_token",
		"refresh_token": s.credentials.RefreshToken,
		"device_code":   s.credentials.DeviceCode,
		"client_id":     "blaxel",
	}

	jsonData, err := json.Marshal(refreshData)
	if err != nil {
		return fmt.Errorf("failed to marshal refresh data: %v", err)
	}

	payload := strings.NewReader(string(jsonData))

	req, _ := http.NewRequest("POST", url, payload)

	req.Header.Add("content-type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to refresh token: %v", err)
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(res.Body)

	var finalizeResponse DeviceLoginFinalizeResponse
	if err := json.Unmarshal(body, &finalizeResponse); err != nil {
		panic(err)
	}
	if finalizeResponse.RefreshToken == "" {
		finalizeResponse.RefreshToken = s.credentials.RefreshToken
	}
	creds := Credentials{
		AccessToken:  finalizeResponse.AccessToken,
		RefreshToken: finalizeResponse.RefreshToken,
		ExpiresIn:    finalizeResponse.ExpiresIn,
		DeviceCode:   s.credentials.DeviceCode,
	}

	s.credentials = creds

	SaveCredentials(s.workspaceName, creds)

	return nil
}
